### PR TITLE
libc: string.h: add cmp functions tests

### PR DESCRIPTION
**Pull request by maska989**
## Description
Added 
 *    memcmp(),
 *    strcmp(),
 *    strncmp(),
 *    strcoll()
 
function tests.

## Motivation and Context
JIRA: CI-231

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [X] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing linter checks and tests passed.
- [X] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: libc/string_cmp.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+{
+}
+
+
+TEST_TEAR_DOWN(string_cmp)
+{
+}
+
+
+TEST(string_cmp, cmp)
```
Comment (damianloew) [inline libc/string_cmp.c:49]: Only a quick look: please split this test to a few cases and make sure that there are no more edge cases/things to check.

**Review by mateusz-bloch — CHANGES_REQUESTED** [review]

File: libc/string_cmp.c [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST(string_cmp, strncmp_basic)
+{
+	TEST_ASSERT_EQUAL_INT(0, strncmp(asciiSet, asciiSet, BUFF_SIZE));
```
[the remark this answers is not in the document]
Comment (maska989) [reply libc/string_cmp.c:106]: It was intentional as you can see in one line after this we used asciiSet and nonZeroAsciiSet.
This case supposes to force *cmp to compare elements that are the same and we try to catch if it will assign data correctly.

File: libc/string_cmp.c [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	RUN_TEST_CASE(string_cmp, strncmp_empty);
+	RUN_TEST_CASE(string_cmp, strncmp_sizes);
+
+	RUN_TEST_CASE(string_cmp, strcmp_basic);
+	RUN_TEST_CASE(string_cmp, strcmp_empty);
+	RUN_TEST_CASE(string_cmp, strcmp_sizes);
+
+	RUN_TEST_CASE(string_cmp, strcoll_basic);
+	RUN_TEST_CASE(string_cmp, strcoll_empty);
+	RUN_TEST_CASE(string_cmp, strcoll_sizes);
```
[the remark this answers is not in the document]
Comment (maska989) [reply libc/string_cmp.c:237]: Those cases check asciiStr where are present null terms.

File: libc/string_cmp.c [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST_SETUP(string_strcoll)
+{
+	errno = 0;
```
Comment (damianloew) [inline libc/string_cmp.c:323]: Why only in tests of strcoll function? Unfortunately it's not certain that errno won't change during the execution of some internal functions:
Screenshot from 2023-06-27 16-03-55

**Review by damianloew — CHANGES_REQUESTED** [review]
Thanks for correcting the test to this version, it looks clear, but let's improve it a little bit.

File: libc/string/string_cmp.c [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+
+		/*
+		 * The testing compares elements that value are every time on a different spot
+		 * and will check if the placement of the dataset on s1 and s2 will not impact on this
+		 */
+		support[i] = 1;
+		TEST_ASSERT_LESS_THAN_INT(0, memcmp(empty, support, BUFF_SIZE));
+		TEST_ASSERT_GREATER_THAN_INT(0, memcmp(support, empty, BUFF_SIZE));
+		support[i] = 0;
+	}
```
Comment (damianloew) [inline libc/string/string_cmp.c:137]: Not tested, but this approach should be more effective.

```suggestion
	/* +1 to provide all ascii characters (ChrStr contains double 1 at the beginning) */
	const char *asciiStr = testdata_createCharStr(BUFF_SIZE + 1); //not sure whether const is possible there
	char asciiStr[BUFF_SIZE +1];
		
	memcpy(asciiStr2, asciiStr, sizeof(asciiStr));

	for (i = 1; i < BUFF_SIZE; i++) {
		asciiStr2[i] = asciiStr[i] - 1;
		TEST_ASSERT_EQUAL_INT(0, memcmp(asciiStr, asciiStr2, i));
		TEST_ASSERT_GREATER_THAN_INT(0, memcmp(asciiStr, asciiStr2, i + 1));
		asciiStr2[i] = asciiStr[i];
	}
	
	free(asciiStr);
	
```

 - you want to test especially cases, where different size changes the result of memcmp, that's why the approach above had been suggested.
 - comment not needed as it should be obvious that we check various sizes in such case,
 - there is no problem to use ascii string with NUL term, the exact data does not matter there, so I'd name those variables asciiStr
 - local variables seems to be better there, because we see the whole flow of data within them.

File: libc/string/string_cmp.c [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+TEST_GROUP(string_strncmp);
+TEST_GROUP(string_strcmp);
+TEST_GROUP(string_strcoll);
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST_SETUP(string_memcmp)
```
Comment (damianloew) [inline libc/string/string_cmp.c:54]: Thanks to the changes below you won't need most of the code there.

File: libc/string/string_cmp.c [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+
+static char empty[BUFF_SIZE] = { 0 },
+			support[BUFF_SIZE] = { 0 },
+			hugeZeroString[BUFF_SIZE * 4] = { 0 },
+			separated[24] = { 0, 0, 0, 0, 0, 0, 'T', 'E', 'S', 'T', 0, 0, 0 },
+			charEdgeValue[1] = { 0 },
+			*asciiSet,
+			*hugeString,
+			*nonZeroAsciiSet,
+			basicStr[6] = "Basic";
```
Comment (damianloew) [inline libc/string/string_cmp.c:42]: Thanks to the suggestions below you shouldn't need most of variables there.

File: libc/string_cmp.c [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+TEST(string_memcmp, sizes)
+{
+	int i;
+
+	/* Memory cmp is not sensitive for NUL characters */
+	TEST_ASSERT_NOT_EQUAL_INT(0, memcmp(empty, separated, sizeof(separated)));
+
+
+	/* Loop which changes the size of compared space to place where asciSet ends */
+	for (i = 1; i < sizeof(asciiSet); i++) {
```
[the remark this answers is not in the document]
Comment (damianloew) [reply libc/string_cmp.c]: It was not applied, I've suggested sth similar, but with the ascii data and with checking two sizes in one iteration.

File: libc/string/string_cmp.c [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+	}
+}
+
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST_SETUP(string_strncmp)
```
Comment (damianloew) [inline libc/string/string_cmp.c:181]: Please apply the suggestions above to the rest of functions too, in case of doubts please let me know.

**Review by damianloew — CHANGES_REQUESTED** [review]

File: libc/string/string_cmp.c [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+TEST(string_memcmp, unsigned_char_cast)
+{
+	/*
+	 * Checking do memcmp convert elements of input to unsigned char
+	 * If we get a value less than 0 that means memcmp treats the value of s1 as
+	 * signed char and return diff between SCHAR_MIN and 0
+	 * otherwise it will calculate the difference between 254 and 0 and memcmp will return
+	 * negative value
+	 */
+	charEdgeValue[0] = SCHAR_MIN;
```
[the remark this answers is not in the document]
Comment (maska989) [reply libc/string/string_cmp.c]: In memcmp it will pass but with str* it throws warnings about the wrong signature that's why i used only char without signed prefix.

File: libc/string/string_cmp.c [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+		 *nonZeroAsciiStr = testdata_createCharStr(BUFF_SIZE);
+	asciiStr[0] = 0;
+
+	TEST_ASSERT_EQUAL_INT(0, strcmp(empty, asciiStr));
+	TEST_ASSERT_EQUAL_INT(0, strcmp(asciiStr, empty));
+	TEST_ASSERT_LESS_THAN_INT(0, strcmp(empty, nonZeroAsciiStr));
+	TEST_ASSERT_GREATER_THAN_INT(0, strcmp(nonZeroAsciiStr, empty));
+
+	free(asciiStr);
+	free(nonZeroAsciiStr);
```
Comment (damianloew) [inline libc/string/string_cmp.c:313]:
```suggestion
	char *asciiStr = testdata_createCharStr(BUFF_SIZE),

	TEST_ASSERT_LESS_THAN_INT(0, strcmp(empty, asciiStr));
	TEST_ASSERT_GREATER_THAN_INT(0, strcmp(asciiStr, empty));

	free(asciiStr);
```

Comment (damianloew) [reply libc/string/string_cmp.c:313]: What about `strcmp(empty, empty)`?

File: libc/string/string_cmp.c [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+	TEST_ASSERT_EQUAL_INT(0, strcoll(hugeStr, hugeStr));
+	/* Comparing the same strings, that are placed in different location */
+	TEST_ASSERT_EQUAL_INT(0, strcoll(hugeStr, hugeStr2));
+	hugeStr[BIG_SIZE - 2] = 1;
+	hugeStr2[sizeof(hugeStr2) - 2] = 2;
+	TEST_ASSERT_LESS_THAN_INT(0, strcoll(hugeStr, hugeStr2));
+	TEST_ASSERT_GREATER_THAN_INT(0, strcoll(hugeStr2, hugeStr));
+	hugeStr[sizeof(hugeStr2) - 2] = 0;
+	TEST_ASSERT_LESS_THAN_INT(0, strcoll(hugeStr, hugeStr2));
+	TEST_ASSERT_GREATER_THAN_INT(0, strcoll(hugeStr2, hugeStr));
```
Comment (damianloew) [inline libc/string/string_cmp.c:396]:
```suggestion
```

**Review by damianloew — CHANGES_REQUESTED** [review]

File: libc/string/string_cmp.c [hunk 13]
```diff
@@ -0,0 +1,439 @@
...
+}
+
+
+TEST(string_strncmp, basic)
+{
+	const char *basicStr = "Basic";
+
+	TEST_ASSERT_EQUAL_INT(0, strncmp(basicStr, basicStr, 6));
+	TEST_ASSERT_GREATER_THAN_INT(0, strncmp(basicStr, support, 6));
+	TEST_ASSERT_LESS_THAN_INT(0, strncmp(support, basicStr, 6));
```
Comment (damianloew) [inline libc/string/string_cmp.c:188]: I think you can also use "Test" or sth like that and we could remove `support` buff.

File: libc/string/string_cmp.c [hunk 14]
```diff
@@ -0,0 +1,439 @@
...
+
+#include "testdata.h"
+
+
+#define BUFF_SIZE 129
+#define BIG_SIZE  1024
+
+static char empty[BUFF_SIZE] = { 0 },
+			support[BUFF_SIZE] = { 0 },
+			separated[24] = { 0, 0, 0, 0, 0, 0, 'T', 'E', 'S', 'T', 0, 0, 0 };
```
Comment (damianloew) [inline libc/string/string_cmp.c:37]: There are only 2/3 calls now, we can safely change them to using "\0\0\0\0\0TEST\0\0".

```suggestion
```

File: libc/string/string_cmp.c [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+#include <unistd.h>
+#include <unity_fixture.h>
+
+#include "testdata.h"
+
+
+#define BUFF_SIZE 129
+#define BIG_SIZE  1024
+
+static char empty[BUFF_SIZE] = { 0 },
```
Comment (damianloew) [inline libc/string/string_cmp.c:35]: It's y default initialized to 0, when it's global buffer.

```suggestion
static char empty[BUFF_SIZE];
```

**Review by damianloew — CHANGES_REQUESTED** [review]

File: libc/string/string_cmp.c [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(&charEdgeValue, "\0", 1));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp("\0", &charEdgeValue, 1));
+}
+
+
+TEST(string_memcmp, emptyInput)
+{
+	char *asciiSet = testdata_createCharStr(BUFF_SIZE),
+		 separated[24] = "\0\0\0\0\0TEST\0\0";
```
Comment (damianloew) [inline libc/string/string_cmp.c:79]: IMHO we should choose one of the following options, because the current one may be misleading:

 - separated[] = "\0\0\0\0\0TEST\0\0\0\0\0\0\0\0\0\0\0\0\0\0"; // sttring with size 24 for init
 - separated[] = "\0\0\0\0\0TEST\0\0"; //current sttring for init
 
I think the second one would be sufficient.

**Review by damianloew — CHANGES_REQUESTED** [review]
Small thing, but I think we should be consistent and start with a lowercase letter after suffix in commit message, so `testdata: extend description for createCharStr`

**Review by damianloew — APPROVED** [review]
LGTM, Let's wait with merge for https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/230

**Review by mateusz-bloch — APPROVED** [review]
